upack/Registry.cs

```diff
@@ -0,0 +1,296 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Reflection;
+using System.Runtime.Serialization;
+using System.Runtime.Serialization.Json;
+using System.Threading.Tasks;
+
+namespace Inedo.ProGet.UPack
+{
+    public sealed class Registry
+    {
+        public static Registry Machine => new Registry(Environment.OSVersion.Platform == PlatformID.Unix ? "/var/lib/upack" : Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData), "upack"));
+        public static Registry User => new Registry(Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".upack"));
+
+        private readonly string path;
+
+        public Registry(string path)
+        {
+            this.path = path;
+        }
+
+        private async Task<T> WithLockAsync<T>(Func<Task<T>> f, string description)
+        {
+            if (description != null && description.Contains("\n"))
+            {
+                throw new ArgumentException("Description must not contain line breaks.");
+            }
+
+            var lockPath = Path.Combine(this.path, ".lock");
+            if (!Directory.Exists(this.path))
+            {
+                Directory.CreateDirectory(path);
+            }
+
+            Stream stream;
+            try
+            {
+                stream = new FileStream(lockPath, FileMode.CreateNew, FileAccess.Write, FileShare.None);
+            }
+            catch (IOException ex)
+            {
+                if (!File.Exists(lockPath))
+                {
+                    throw;
+                }
+
+                var lastWrite = File.GetLastWriteTime(lockPath);
+                if (lastWrite + TimeSpan.FromSeconds(10) < DateTime.Now)
+                {
+                    stream = new FileStream(lockPath, FileMode.OpenOrCreate, FileAccess.ReadWrite, FileShare.None);
+                    if (File.GetLastWriteTime(lockPath) != lastWrite && stream.Length != 0)
+                    {
+                        string lockDescription = null;
+
+                        try
+                        {
+                            using (var reader = new StreamReader(stream))
+                            {
+                                lockDescription = await reader.ReadLineAsync().ConfigureAwait(false);
+                            }
+                        }
+                        catch
+                        {
+                        }
+
+                        throw new RegistryLockedException("Registry is locked: " + (lockDescription ?? "No description provided."), ex);
```

Comment:
Is this going to be automatically retried with a catch somewhere else? I didn't see it... 

Anyway we shouldn't just throw an exception in this case, but retry (up to 1000 times) after 3 seconds.